src/actions/sagemaker/sagemaker_infer.ts

```diff
@@ -0,0 +1,356 @@
+      logJson("transformResponse", transformResponse)
+
+      // return success response
+      return new Hub.ActionResponse({ success: true })
+
+    } catch (err) {
+      winston.error(err)
+      return new Hub.ActionResponse({ success: false, message: err.message })
+    }
+  }
+
+  async form(request: Hub.ActionRequest) {
+    winston.debug("form")
+    logJson("request.params", request.params)
+
+    const buckets = await this.listBuckets(request)
+    if (! Array.isArray(buckets)) {
+      throw "Unable to retrieve buckets"
+    }
+    logJson("buckets", buckets)
+
+    const models = await this.listModels(request)
+    if (! Array.isArray(models)) {
+      throw "Unable to retrieve models"
+    }
+    logJson("models", models)
+
+    const form = new Hub.ActionForm()
+    form.fields = [
+      {
+        label: "Model",
+        name: "modelName",
+        required: true,
+        options: models.map((model: any) => {
+          return {
+            name: model.ModelName,
+            label: model.ModelName,
+          }
+        }),
+        type: "select",
+        description: "The S3 bucket where SageMaker input training data should be stored",
+      },
+      {
+        label: "Strip Columns",
+        name: "numStripColumns",
+        required: true,
+        options: [
+          { name: "0", label: "None" },
+          { name: "1", label: "First Column" },
+          { name: "2", label: "First & Second Column" },
+        ],
+        type: "select",
+        default: "0",
+        description: "Columns to remove before running inference task. Columns must be first or second column in the data provided. Use this to remove key, target variable, or both.",
+      },
+      {
+        label: "Output Bucket",
+        name: "bucket",
+        required: true,
+        options: buckets.map((bucket) => {
+          return {
+            name: bucket.Name!,
+            label: bucket.Name!,
+          }
+        }),
+        type: "select",
+        // default: buckets[0].Name, // DNR
+        default: "looker-marketing-analysis", // DNR
+        description: "The S3 bucket where inference data should be stored",
+      },
+      {
+        type: "select",
+        label: "AWS Instance Type",
+        name: "awsInstanceType",
+        required: true,
+        options: awsInstanceTypes.map((type) => {
+          return {
+            name: type,
+            label: type,
+          }
+        }),
+        default: "ml.m4.xlarge",
+        description: "The type of AWS instance to use. More info: More info: https://aws.amazon.com/sagemaker/pricing/instance-types",
+      },
+      {
+        type: "string",
+        label: "Number of instances",
+        name: "numInstances",
+        default: "1",
+        description: "The number of instances to run. Valid values: 1 to 500.",
+      },
+    ]
+    return form
+  }
+
+  protected getSageMakerClientFromRequest(request: Hub.ActionRequest) {
```

Comment:
I'm still familiarizing with TypeScript, is it typical to allow type inference for return types when we're sure we won't get a different result (i.e. `undefined`)?

Reply:
yeah i'm still learning it myself, but I think TS is able to tell that this function always returns a SageMaker instance so no need to declare that in the function signature.